constraint/pkg/client/client.go

```diff
@@ -135,29 +166,72 @@ func (c *Client) AddTemplate(ctx context.Context, templ *templates.ConstraintTem
 		return resp, err
 	}
 
-	if err := c.driver.AddTemplate(ctx, templ); err != nil {
+	newDriverN := c.driverForTemplate(templ)
+
+	driver, ok := c.drivers[newDriverN]
+	if !ok {
+		return resp, fmt.Errorf("%w: available drivers: %v, wanted %q", clienterrors.ErrNoDriver, c.driverPriority, c.driverForTemplate(templ))
+	}
+
+	// TODO: because different targets may have different code sets,
+	// the driver should be told which targets to load code for.
+	// this is moot right now, since templates only have one target
+	if err := driver.AddTemplate(ctx, templ); err != nil {
 		return resp, err
 	}
 
 	templateName := templ.GetName()
 
-	template := c.templates[templateName]
+	cacheEntry := c.templates[templateName]
 
 	// We don't want to use the usual "if found/ok" idiom here - if the value
 	// stored for templateName is nil, we need to update it to be non-nil to avoid
 	// a panic.
-	if template == nil {
-		template = &templateClient{
-			constraints: make(map[string]*constraintClient),
+	if cacheEntry == nil {
+		cacheEntry = newTemplateClient()
+		c.templates[templateName] = cacheEntry
+	}
+
+	cacheEntry.activeDrivers[newDriverN] = true
+
+	// For drivers that require a local cache of constraints, we ensure that
+	// cache is current if the active driver has changed.
+	if cachedCpy != nil {
+		oldDriverN := c.driverForTemplate(cachedCpy)
+		if oldDriverN != newDriverN {
+			cacheEntry.needsConstraintReplay = true
 		}
+	}
 
-		c.templates[templateName] = template
+	if cacheEntry.needsConstraintReplay {
+		for _, constraintEntry := range cacheEntry.constraints {
+			cstr := constraintEntry.getConstraint()
+			if err := driver.AddConstraint(ctx, cstr); err != nil {
+				return resp, fmt.Errorf("%w: while replaying constraints", err)
+			}
+		}
+		cacheEntry.needsConstraintReplay = false
 	}
 
-	// This state mutation needs to happen last so that the semantic equal check
-	// at the beginning does not incorrectly return true when updating did not
-	// succeed previously.
-	template.Update(templ, crd, target)
+	// This state mutation needs to happen after the new driver is fully ready
+	// to enforce the template
+	cacheEntry.Update(templ, crd, target)
+
+	// Remove old drivers last so that templates can be enforced
+	// despite a botched update
+	for oldDriverN := range cacheEntry.activeDrivers {
+		if oldDriverN == newDriverN {
+			continue
+		}
+		oldDriver, ok := c.drivers[oldDriverN]
+		if !ok {
+			return resp, fmt.Errorf("%w: while changing drivers", clienterrors.ErrNoDriver)
+		}
```

Comment:
Failing early is simpler, since there would only be one error to return. Continuing iteration means risking adding more errors.

Because the recovery modes here are:

* keep retrying and hope the failure goes away
* destroy the client and rebuild from scratch

there is little functional difference between the two approaches.

You could argue that it's then possible for every registered driver to have stale state in very specific circumstances if neither of the recovery modes are satisfied, but in practice the maximum # of drivers are hardcoded, and if this mode is common, there is likely a larger issue.

Reply:
Also, if this winds up being the wrong choice, it can be changed.